lib/opentelemetry_plug.ex

```diff
@@ -82,47 +78,64 @@ defmodule OpentelemetryPlug do
       {"http.target", conn.request_path},
       {"http.host", conn.host},
       {"http.scheme", conn.scheme},
+      {"http.flavor", http_flavor(conn.adapter)},
+      {"http.route", route},
       {"http.user_agent", user_agent},
       {"http.method", conn.method},
       {"net.peer.ip", to_string(:inet_parse.ntoa(peer_ip))},
       {"net.peer.port", peer_data.port},
       {"net.peer.name", host},
       {"net.transport", "IP.TCP"},
       {"net.host.ip", to_string(:inet_parse.ntoa(conn.remote_ip))},
-      {"net.host.port", conn.port} | optional_attributes(conn)
-      # {"net.host.name", HostName}
+      {"net.host.port", conn.port}
+      | optional_attributes(conn)
     ]
 
     # TODO: Plug should provide a monotonic native time in measurements to use here
     # for the `start_time` option
-    OpenTelemetry.Tracer.start_span(span_name, %{attributes: attributes})
-  end
+    span_ctx = Tracer.start_span(span_name, %{attributes: attributes, kind: :server})
 
-  @doc false
-  def handle_route(_, _measurements, %{route: route}, _config) do
-    # TODO: add config option to allow `conn.request_path` as span name
-    if in_span?() do
-      OpenTelemetry.Span.update_name(route)
-    end
+    Tracer.set_current_span(span_ctx)
   end
 
   @doc false
   def handle_stop(_, _measurements, %{conn: conn}, _config) do
     if in_span?() do
-      OpenTelemetry.Span.set_attribute("http.status", conn.status)
-      OpenTelemetry.Tracer.end_span()
+      Tracer.set_attribute("http.status_code", conn.status)
+      # For HTTP status codes in the 4xx and 5xx ranges, as well as any other
+      # code the client failed to interpret, status MUST be set to Error.
+      #
+      # Don't set the span status description if the reason can be inferred from
+      # http.status_code.
+      if conn.status >= 400 do
+        Tracer.set_status(OpenTelemetry.status(:error, ""))
+      end
+
+      Tracer.end_span()
+      restore_parent_ctx()
     end
   end
 
   @doc false
-  def handle_exception(_, _measurements, %{conn: _conn}, _config) do
+  def handle_exception(_, _measurements, metadata, _config) do
     if in_span?() do
-      OpenTelemetry.Span.set_status(OpenTelemetry.status('UnknownError', "unknown error"))
-      OpenTelemetry.Tracer.end_span()
+      %{kind: kind, reason: reason, stacktrace: stacktrace} = metadata
+      exception = Exception.normalize(kind, reason, stacktrace)
+
+      Span.record_exception(
+        Tracer.current_span_ctx(),
+        exception,
+        stacktrace
+      )
+
+      Tracer.set_status(OpenTelemetry.status(:error, Exception.message(exception)))
+      Tracer.set_attribute("http.status_code", 500)
+      Tracer.end_span()
+      restore_parent_ctx()
```

Comment:
You shouldn't need to set this. The plug will continue to run through the stop event where the end error code is already set. Also, the stop event would be where the span is getting closed, so you I don't think you should be ending the span or restoring the parent.

The flow (at least in Phoenix) is always one of these two:

* `start -> stop`
* `start -> exception -> stop`

Reply:
It appears to my eyes that on [`Plug.Router` it's always](https://github.com/elixir-plug/plug/blob/master/lib/plug/router.ex#L283-L303):
* `start -> stop`
* `start -> exception`

The `else` clause is never called after `catch`, so the `stop` event won't be invoked in that path.

I'll remove the context-restoring stuff.

Reply:
Removed the context save/restore in 92b6304